### ISSUE TITLE
Can't build out-of-the-box on pre-8.4 GHCs

### ISSUE THREAD

**Issue by quasicomputational** [haskell/cabal#5422]
In `Cabal.cabal`, we have this unconditional `build-depends` delcaration:

```
  -- Parsec parser-related modules
  build-depends:
    -- transformers-0.4.0.0 doesn't have record syntax e.g. for Identity
    -- See also https://github.com/ekmett/transformers-compat/issues/35
    transformers (>= 0.3      && < 0.4) || (>=0.4.1.0 && <0.6),
    mtl           >= 2.1      && < 2.3,
    text          >= 1.2.3.0  && < 1.3,
    parsec        >= 3.1.13.0 && < 3.2
```

But none of those are bundled with GHC before 8.4, which means that our support promise ("the Cabal library must be buildable out-of-the-box with the dependencies that shipped with GHC for at least five years") isn't being fulfilled.

**Comment by hvr** [comment]
@quasicomputational yeah, that was unavoidable once we started depending on `parsec`; I think we discussed this somewhere but I can't find it right now. It's also worth revisiting what the purpose of the "out-of-box buildability" desiderata is.

/cc @phadej
